src/level/level_loader.cpp

```diff
@@ -0,0 +1,1201 @@
+
+            switch (m_main->m_levelCategory)
+            {
+            case LevelCategory::Missions:
+                m_main->m_playerProfile->SetFreeGameResearchUnlock(m_main->m_playerProfile->GetFreeGameResearchUnlock() | m_main->m_researchDone[0]);
+                m_main->m_playerProfile->SetFreeGameBuildUnlock(m_main->m_playerProfile->GetFreeGameBuildUnlock() | m_main->m_build);
+                break;
+            case LevelCategory::FreeGame:
+                m_main->m_researchDone[0] = m_main->m_playerProfile->GetFreeGameResearchUnlock();
+                m_main->m_build = m_main->m_playerProfile->GetFreeGameBuildUnlock();
+                m_main->m_build &= ~BUILD_RESEARCH;
+                m_main->m_build &= ~BUILD_LABO;
+                m_main->m_build |= BUILD_FACTORY;
+                m_main->m_build |= BUILD_GFLAT;
+                m_main->m_build |= BUILD_FLAG;
+                break;
+            default:
+                break;
+            }
+
+            m_main->m_short->SetMode(false);  // vehicles?
+        }
+
+        m_main->m_map->ShowMap(m_main->m_mapShow);
+        m_main->m_map->UpdateMap();
+        // TODO: m_main->m_engine->TimeInit(); ??
+        m_main->m_input->ResetKeyStates();
+        m_main->m_time = 0.0f;
+        m_main->m_gameTime = 0.0f;
+        m_main->m_gameTimeAbsolute = 0.0f;
+        m_main->m_autosaveLast = 0.0f;
+        m_main->m_infoUsed = 0;
+
+        m_main->m_selectObject = m_selectedObject;
+
+        if (m_main->m_base == nullptr &&  // no m_main base?
+            !m_main->m_fixScene)          // interractive scene?
+        {
+            CObject* obj = m_selectedObject;
+            if (m_selectedObject == nullptr)
+                obj = m_main->SearchHuman();
+
+            if (obj != nullptr)
+            {
+                assert(obj->Implements(ObjectInterfaceType::Controllable));
+                m_main->SelectObject(obj);
+                m_main->m_camera->SetControllingObject(obj);
+                m_main->m_camera->SetType(dynamic_cast<CControllableObject*>(obj)->GetCameraType());
+            }
+        }
+
+        if (m_main->m_fixScene)
+            m_main->m_camera->SetType(Gfx::CAM_TYPE_SCRIPT);
+
+        if (mode == LevelLoadMode::LoadSavedGame && m_selectedObject != nullptr)  // loading file?
+        {
+            Math::Vector pos = m_selectedObject->GetPosition();
+            m_main->m_camera->Init(pos, pos, 0.0f);
+
+            m_main->SelectObject(m_selectedObject);
+            m_main->m_camera->SetControllingObject(m_selectedObject);
+
+            m_main->m_beginSatCom = true;  // message already displayed
+        }
+    }
+    catch (...)
+    {
+        m_main->m_sceneReadPath = "";
+        throw;
+    }
+    m_main->m_sceneReadPath = "";
+
+    if (m_main->m_app->GetSceneTestMode())
+        m_main->m_eventQueue->AddEvent(Event(EVENT_QUIT));
+
+    m_main->m_ui->ShowLoadingScreen(false);
+    if (m_main->m_missionType == MISSION_CODE_BATTLE)
```

Comment:
If you are going to clean up level loading, do it right. Creating the UI really shouldn't be invoked from CreateScene, CRobotMain::ChangePhase would probably be a better place.